File: src/examples/sensors_tutorial.cpp
#include <robot_dart/control/pd_control.hpp>
#include <robot_dart/robot_dart_simu.hpp>
#include <robot_dart/robots/iiwa.hpp>

#include <robot_dart/gui/magnum/graphics.hpp>
#include <robot_dart/gui/magnum/sensor/camera.hpp>
#include <robot_dart/sensor/force_torque.hpp>
#include <robot_dart/sensor/imu.hpp>
#include <robot_dart/sensor/torque.hpp>

int main()
{
    auto robot = std::make_shared<robot_dart::robots::Iiwa>();
    robot->set_actuator_types("torque");

    Eigen::VectorXd ctrl = robot_dart::make_vector({0., M_PI / 3., 0., -M_PI / 4., 0., 0., 0.});

    auto controller = std::make_shared<robot_dart::control::PDControl>(ctrl);
    robot->add_controller(controller);
    controller->set_pd(300., 50.);

    robot_dart::RobotDARTSimu simu(0.001);

    auto graphics = std::make_shared<robot_dart::gui::magnum::Graphics>();
    simu.set_graphics(graphics);

    graphics->look_at({0., 3.5, 2.}, {0., 0., 0.25});

    simu.add_robot(robot);
    simu.add_checkerboard_floor();
    Eigen::Vector6d pose;
    pose << 0., 0., 0., 1.5, 0., 0.25;

    simu.add_robot(robot_dart::Robot::create_box({0.1, 0.1, 0.5}, pose, "free", 1., dart::Color::Red(1.), "box"));
    // Format Eigen to std::cout
    Eigen::IOFormat fmt(Eigen::StreamPrecision, Eigen::DontAlignCols, " ", "\n", "", "");
    std::cout.precision(5);

    robot->set_damping_coeffs(0.);
    robot->set_coulomb_coeffs(0.);
    robot->set_spring_stiffnesses(0.);

    auto dampings = robot->damping_coeffs();
    auto frictions = robot->coulomb_coeffs();
    auto stiffnesses = robot->spring_stiffnesses();

    for (size_t i = 0; i < dampings.size(); ++i) {
        std::cout << "DoF #" << i << std::endl;
        std::cout << "    Stiffness: " << stiffnesses[i] << std::endl;
        std::cout << "    Friction: " << frictions[i] << std::endl;
        std::cout << "    Damping: " << dampings[i] << std::endl;
    }
    auto camera = std::make_shared<robot_dart::sensor::Camera>(graphics->magnum_app(), 256, 256);

    camera->camera().record(true, true); // cameras are recording color images by default, enable depth images as well for this example
    // cameras can also record video
    camera->record_video("video-camera.mp4");

    // attach camera to body
    Eigen::Isometry3d tf;
    tf = Eigen::AngleAxisd(3.14, Eigen::Vector3d{1., 0., 0.});
    tf *= Eigen::AngleAxisd(1.57, Eigen::Vector3d{0., 0., 1.});
    tf.translation() = Eigen::Vector3d(0., 0., 0.1);
    camera->attach_to_body(robot->body_node("iiwa_link_ee"), tf); // cameras are looking towards -z by default

    // @TORQUE_SENSOR@
    // Add torque sensors to the robot
    int ct = 0;
    std::vector<std::shared_ptr<robot_dart::sensor::Torque>> tq_sensors(robot->num_dofs());
    for (const auto& joint : robot->dof_names())
        tq_sensors[ct++] = simu.add_sensor<robot_dart::sensor::Torque>(robot, joint, 1000);
    // @TORQUE_SENSOR_END@

    // @FORCE_TORQUE_SENSOR@
    // Add force-torque sensors to the robot
    ct = 0;
    std::vector<std::shared_ptr<robot_dart::sensor::ForceTorque>> f_tq_sensors(robot->num_dofs());
    for (const auto& joint : robot->dof_names())
        f_tq_sensors[ct++] = simu.add_sensor<robot_dart::sensor::ForceTorque>(robot, joint, 1000, "parent_to_child");
    // @TORQUE_FORCE_SENSOR_END@

    // @IMU_SENSOR@
    // Add IMU sensors to the robot
    ct = 0;
    std::vector<std::shared_ptr<robot_dart::sensor::IMU>> imu_sensors(robot->num_bodies());
    for (const auto& body_node : robot->body_names()) {
        // _imu(std::make_shared<sensor::IMU>(sensor::IMUConfig(body_node("head"), frequency))),
        imu_sensors[ct++] = simu.add_sensor<robot_dart::sensor::IMU>(robot_dart::sensor::IMUConfig(robot->body_node(body_node), 1000));
    }
    // @IMU_SENSOR_END@

    // Cameras can be used as sensors to
    simu.add_sensor(camera);

    auto start = std::chrono::steady_clock::now();
    Eigen::Vector3d external_force = Eigen::Vector3d::Zero();

    while (simu.scheduler().next_time() < 20 && !simu.graphics()->done()) {
        simu.step();

        // Print torque sensor measurement
        if (simu.schedule(tq_sensors[0]->frequency())) {
            ct = 0;
            // @TORQUE_MEASUREMENT@
            // vector that contains the torque measurement for every joint (scalar)
            Eigen::VectorXd torques_measure(robot->num_dofs());
            for (const auto& tq_sens : tq_sensors)
                torques_measure.block<1, 1>(ct++, 0) = tq_sens->torques();
            // @TORQUE_MEASUREMENT_END@
            // @FORCE_TORQUE_MEASUREMENT@
            //  matrix that contains the torque measurement for every joint (3d vector)
            Eigen::MatrixXd ft_torques_measure(robot->num_dofs(), 3);
            //  matrix that contains the force measurement for every joint (3d vector)
            Eigen::MatrixXd ft_forces_measure(robot->num_dofs(), 3);
            //  matrix that contains the wrench measurement for every joint (6d vector)[torque, force]
            Eigen::MatrixXd ft_wrench_measure(robot->num_dofs(), 6);
            ct = 0;
            for (const auto& f_tq_sens : f_tq_sensors) {
                ft_torques_measure.block<1, 3>(ct, 0) = f_tq_sens->torque();
                ft_forces_measure.block<1, 3>(ct, 0) = f_tq_sens->force();
                ft_wrench_measure.block<1, 6>(ct, 0) = f_tq_sens->wrench();
                ct++;
            }
            // @FORCE_TORQUE_MEASUREMENT_END@

            // @IMU_MEASUREMENT@
            Eigen::MatrixXd imu_angular_positions_measure(robot->num_bodies(), 3);
            Eigen::MatrixXd imu_angular_velocities_measure(robot->num_bodies(), 3);
            Eigen::MatrixXd imu_linear_acceleration_measure(robot->num_bodies(), 3);
            ct = 0;
            for (const auto& imu_sens : imu_sensors) {
                imu_angular_positions_measure.block<1, 3>(ct, 0) = imu_sens->angular_position_vec();
                imu_angular_velocities_measure.block<1, 3>(ct, 0) = imu_sens->angular_velocity();
                imu_linear_acceleration_measure.block<1, 3>(ct, 0) = imu_sens->linear_acceleration();
                ct++;
            }
            // @IMU_MEASUREMENT_END@

            std::cout
                << " Torque sensors' torque:  \n"
                << torques_measure.transpose().format(fmt) << std::endl;
            std::cout << " Force-Torque sensors' torque:  \n"
                      << ft_torques_measure.transpose().format(fmt) << std::endl;
            std::cout << " Force-Torque sensors' force:  \n"
                      << ft_forces_measure.format(fmt) << std::endl;
            std::cout << " Force-Torque sensors' wrench:  \n"
                      << ft_wrench_measure.format(fmt) << std::endl;
            std::cout << "IMU sensors Angular Position: \n"
                      << imu_angular_positions_measure.format(fmt) << std::endl;
            std::cout << "IMU sensors Angular Velocity: \n"
                      << imu_angular_positions_measure.format(fmt) << std::endl;
            std::cout << "IMU sensors Linear Acceleration: \n"
                      << imu_angular_positions_measure.format(fmt) << std::endl;
            std::cout << "=================================" << std::endl;
        }

        // add external force
        int time_in_seconds = static_cast<int>(simu.scheduler().next_time());
        if (time_in_seconds % 3 >= 1 && time_in_seconds % 3 < 2) {
            external_force = Eigen::Vector3d::Constant(30.0);
            //std::cout << "Applying force on iiwa_link_4" << std::endl;
        }
        else {

            external_force = Eigen::Vector3d::Zero();
        }

        robot->set_external_force("iiwa_link_4", external_force);
    }
    // @RGB_SENSOR@
    // a nested std::vector (w*h*3) of the last image taken can be retrieved
    auto rgb_image = camera->image();
    // @RGB_SENSOR_END@
    // @RGB_SENSOR_MEASURE@
    // we can also save them to png
    robot_dart::gui::save_png_image("camera-small.png", rgb_image);
    // convert an rgb image to grayscale (useful in some cases)
    auto gray_image = robot_dart::gui::convert_rgb_to_grayscale(rgb_image);
    robot_dart::gui::save_png_image("camera-gray.png", gray_image);
    // @RGB_SENSOR_MEASURE_END@

    // @RGB_D_SENSOR@
    auto rgb_d_image = camera->depth_image();
    // @RGB_D_SENSOR_END@
    // @RGB_D_SENSOR_MEASURE@
    // get the depth image from a camera
    // with a version for visualization or bigger differences in the output
    robot_dart::gui::save_png_image("camera-depth.png", rgb_d_image);
    // and the raw values that can be used along with the camera parameters to transform the image to point-cloud
    robot_dart::gui::save_png_image("camera-depth-raw.png", rgb_d_image);
    // @RGB_D_SENSOR_MEASURE_END@
    auto end = std::chrono::steady_clock::now();
    std::chrono::duration<double> elapsed_seconds = end - start;
    std::cout << "benchmark time: " << elapsed_seconds.count() << "s\n";

    robot.reset();
    return 0;
}

File: src/examples/python/sensors_tutorial.py
import RobotDART as rd
import numpy as np
import magnum
import dartpy
import time
robot = rd.Iiwa()
robot.set_actuator_types("torque")

ctrl = [0., np.pi / 3., 0., -np.pi / 4., 0., 0., 0.]

controller = rd.PDControl(ctrl)
robot.add_controller(controller)
controller.set_pd(300., 50.)

simu = rd.RobotDARTSimu(0.001)

graphics = rd.gui.Graphics()
simu.set_graphics(graphics)

graphics.look_at([0., 3.5, 2.], [0., 0., 0.25])

simu.add_robot(robot)
simu.add_checkerboard_floor()

pose = [0., 0., 0., 1.5, 0., 0.25]

simu.add_robot(rd.Robot.create_box(
    [0.1, 0.1, 0.5], pose, "free", 1., magnum.Color4.red(1.), "box"))


robot.set_damping_coeffs(0.)
robot.set_coulomb_coeffs(0.)
robot.set_spring_stiffnesses(0.)

dampings = robot.damping_coeffs()
frictions = robot.coulomb_coeffs()
stiffnesses = robot.spring_stiffnesses()

for i in range(len(dampings)):
    print("DoF #", i)
    print("    Stiffness: ", stiffnesses[i])
    print("    Friction: ", frictions[i])
    print("    Damping: ", dampings[i])

camera = rd.sensor.Camera(graphics.magnum_app(), 256, 256)

# @CAMERA_SENSOR_RGBD_RECORD_DEPTH_PYTHON@
# cameras are recording color images by default, enable depth images as well for this example
camera.camera().record(True, True)
# @CAMERA_SENSOR_RGBD_RECORD_DEPTH_PYTHON_END@
# cameras can also record video
camera.record_video("video-camera.mp4")

# attach camera to body
tf = dartpy.math.Isometry3()
rot = dartpy.math.AngleAxis(3.14, [1., 0., 0.])
rot = rot.multiply(dartpy.math.AngleAxis(1.57, [0., 0., 1.])).to_rotation_matrix()
tf.set_translation([0., 0., 0.1])
# cameras are looking towards -z by default
camera.attach_to_body(robot.body_node("iiwa_link_ee"), tf)

# @TORQUE_SENSOR_PYTHON@
# Add torque sensors to the robot
tq_sensors = np.empty(robot.num_dofs(), dtype=rd.sensor.Torque)
ct = 0
for joint in robot.dof_names():
    simu.add_sensor(rd.sensor.Torque(robot, joint, 1000))
    tq_sensors[ct] = simu.sensors()[-1]
    ct += 1
# @TORQUE_SENSOR_PYTHON_END@

# @FORCE_TORQUE_SENSOR_PYTHON@
# Add force-torque sensors to the robot
f_tq_sensors = np.empty(robot.num_dofs(), dtype=rd.sensor.ForceTorque)
ct = 0
for joint in robot.dof_names():
    simu.add_sensor(rd.sensor.ForceTorque(
        robot, joint, 1000, "parent_to_child"))
    f_tq_sensors[ct] = simu.sensors()[-1]
    print(f_tq_sensors)
    ct += 1
# @FORCE_TORQUE_SENSOR_PYTHON_END@

# @IMU_SENSOR_PYTHON@
# Add IMU sensors to the robot
ct = 0
imu_sensors = np.empty(robot.num_bodies(), dtype=rd.sensor.IMU)
for body_node in robot.body_names():
    simu.add_sensor(rd.sensor.IMU(
        rd.sensor.IMUConfig(robot.body_node(body_node), 1000)))
    imu_sensors[ct] = simu.sensors()[-1]
    ct += 1
# @IMU_SENSOR_PYTHON_END@

# Cameras can be used as sensors too
simu.add_sensor(camera)
start = time.time()
external_force = np.zeros(3)

while (simu.scheduler().next_time() < 20 and not simu.graphics().done()):
    simu.step()

    # Print torque sensor measurement
    if (simu.schedule(tq_sensors[0].frequency())):
        
        # @TORQUE_MEASUREMENT_PYTHON@
        # vector that contains the torque measurement for every joint (scalar)
        torques_measure = np.empty(robot.num_dofs())
        ct = 0
        for tq_sens in tq_sensors:
            torques_measure[ct] = tq_sens.torques()
            ct += 1
        # @TORQUE_MEASUREMENT_PYTHON_END@
        
        # @FORCE_TORQUE_MEASUREMENT_PYTHON@
        #  matrix that contains the torque measurement for every joint (3d vector)
        ft_torques_measure = np.empty([robot.num_dofs(), 3])
        #  matrix that contains the force measurement for every joint (3d vector)
        ft_forces_measure = np.empty([robot.num_dofs(), 3])
        #  matrix that contains the wrench measurement for every joint (6d vector)[torque, force]
        ft_wrench_measure = np.empty([robot.num_dofs(), 6])
        ct = 0
        for f_tq_sens in f_tq_sensors:
            ft_torques_measure[ct, :] = f_tq_sens.torque()
            ft_forces_measure[ct, :] = f_tq_sens.force()
            ft_wrench_measure[ct, :] = f_tq_sens.wrench()
            ct += 1

        # @FORCE_TORQUE_MEASUREMENT_PYTHON_END@

        # @IMU_MEASUREMENT_PYTHON@
        imu_angular_positions_measure = np.empty([robot.num_bodies(), 3])
        imu_angular_velocities_measure = np.empty([robot.num_bodies(), 3])
        imu_linear_acceleration_measure = np.empty([robot.num_bodies(), 3])
        ct = 0
        for imu_sens in imu_sensors:
            imu_angular_positions_measure[ct,:] = imu_sens.angular_position_vec()
            imu_angular_velocities_measure[ct, :] = imu_sens.angular_velocity()
            imu_linear_acceleration_measure[ct,:] = imu_sens.linear_acceleration()
            ct += 1

        # @IMU_MEASUREMENT_PYTHON_END@

        print("Torque sensors torque: \n ", torques_measure.transpose())
        print("Force-Torque sensors torque: \n ",
              ft_torques_measure.transpose())
        print("Force-Torque sensors force: \n ", ft_forces_measure.transpose())
        print("Force-Torque sensors wrench: \n ",
              ft_wrench_measure.transpose())
        print("IMU sensors Angular Position: \n ",
              imu_angular_positions_measure)
        print("IMU sensors Angular Velocity: \n ",
              imu_angular_velocities_measure)
        print("IMU sensors Linear Acceleration Velocity: \n ",
              imu_linear_acceleration_measure)
        print("=================================")

    # add external force
    time_in_seconds = int(simu.scheduler().next_time())
    if ((time_in_seconds % 3 >= 1) & (time_in_seconds % 3 < 2)):
        external_force = np.ones([3]) * 30.0
        print("Applying force on iiwa_link_4")

    else:
        external_force = np.zeros([3])

    robot.set_external_force("iiwa_link_4", external_force)

# @RGB_SENSOR_PYTHON@
# a nested array (w*h*3) of the last image taken can be retrieved
rgb_image = camera.image()
# @RGB_SENSOR_PYTHON_END@
# @RGB_SENSOR_MEASURE_PYTHON@
# we can also save them to png
rd.gui.save_png_image("camera-small.png", rgb_image)
# convert an rgb image to grayscale (useful in some cases)
gray_image = rd.gui.convert_rgb_to_grayscale(rgb_image)
rd.gui.save_png_image("camera-gray.png", gray_image)
# @RGB_SENSOR_MEASURE_PYTHON_END@

# @RGB_D_SENSOR_PYTHON@
# get the depth image from a camera
# with a version for visualization or bigger differences in the output
rgb_d_image = camera.depth_image()
# and the raw values that can be used along with the camera parameters to transform the image to point-cloud
rgb_d_image_raw = camera.raw_depth_image()
# @RGB_D_SENSOR_PYTHON_END@
# @RGB_D_SENSOR_MEASURE_PYTHON@
rd.gui.save_png_image("camera-depth.png", rgb_d_image)
rd.gui.save_png_image("camera-depth-raw.png", rgb_d_image_raw)
# @RGB_D_SENSOR_MEASURE_PYTHON_END@

end = time.time()
elapsed_seconds = end - start
print("benchmark time: ", elapsed_seconds, "s\n")

robot.reset()
